File: JavaScript/06_ZigzagConversion.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
const convert = function(s, numRows) {
    if (numRows === 1) {
        return s;
    }

    const hashMap = {}
    let zig = true;
    let zag = false;
    let row = 0;
    let result = "";

    for (let i = 0; i < numRows; i++) {
        hashMap[i] = [];
    }

    for (let i = 0; i < s.length; i++) {
        if (zig) {
            hashMap[row].push(s[i]);
            if (row === numRows - 1) {
                zig = false;
                zag = true;
                row--;
            } else {
                row++;
            }
        } else if (zag) {
            hashMap[row].push(s[i])
            if (row === 0) {
                zig = true;
                zag = false;
                row++;
            } else {
                row--; 
            }
        }
    }

    for (let i = 0; i < numRows; i++) {
        hashMap[i].forEach(char => {
            result = result.concat('', char);
        })
    }
    
    return result;
};


testCases = [
    {
        s: "PAYPALISHIRING",
        numRows: 3,
        expected: "PAHNAPLSIIGYIR"
    },
    {
        s: "PAYPALISHIRING",
        numRows: 4,
        expected: "PINALSIGYAHRPI"
    },
    {
        s: "A",
        numRows: 1,
        expected: "A"
    },
    {
        s: "AB",
        numRows: 1,
        expected: "AB"
    },
    {
        s: "AB",
        numRows: 2,
        expected: "AB"
    },
]

testCases.forEach((testCase, i) => {
    const output = convert(testCase.s, testCase.numRows);
    if (JSON.stringify(output) === JSON.stringify(testCase.expected)) {
        console.log(`Test Case ${i}: Passed. Expected: ${JSON.stringify(testCase.expected)}, Output: ${JSON.stringify(output)} `)
    } else {
        console.log(`Test Case ${i}: Failed. Expected: ${JSON.stringify(testCase.expected)}, Output: ${JSON.stringify(output)} `)
    }
    
});

File: Python/001-999/001-099/06_ZigzagConversion.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1:
            return s

        hash_map = {}
        zig = True
        zag = False
        row = 0
        result = ""

        for i in range(numRows):
            hash_map[i] = []

        for char in s:
            if zig:
                hash_map[row].append(char)
                if row == numRows - 1:
                    zig = False
                    zag = True
                    row -= 1
                else:
                    row += 1

            elif zag:
                hash_map[row].append(char)
                if row == 0:
                    zig = True
                    zag = False
                    row += 1

                else:                    
                    row -= 1

        
        for i in range(numRows):
            for char in hash_map[i]:
                result += char
        
        return result
            

s = Solution()

print(s.convert("PAYPALISHIRING", 3))
print(s.convert("PAYPALISHIRING", 4))
print(s.convert("A", 1))
print(s.convert("AB", 1))